Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,91 @@
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return candidate.absoluteBoundingBoxRect().location() - m_frameView.layoutViewportRect().location();
+}
+
+void ScrollAnchoringController::chooseAnchorElement()
+{
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    // https://drafts.csswg.org/css-scroll-anchoring/#anchor-node-selection
+
+    auto* document = m_frameView.frame().document();
+    if (!document)
+        return;
+
+    Element* candidateAnchor = nullptr;
+    auto layoutViewport = m_frameView.layoutViewportRect();
+    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
+        if (auto renderer = element.renderBox()) {
+            if (renderer->style().overflowAnchor() == OverflowAnchor::None || &element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
+                continue;
+            auto boxRect = renderer->absoluteBoundingBoxRect();
+            if (layoutViewport.intersects(boxRect))
+                candidateAnchor = &element;
+            if (layoutViewport.contains(boxRect))
+                break;
+        }
+    }
+    if (!candidateAnchor)
+        return;
+    m_anchorElement = candidateAnchor;
+    m_lastOffsetForAnchorElement = computeOffset(*m_anchorElement->renderBox());
+}
+
+void ScrollAnchoringController::updateAnchorElement()
+{
+    if (frameView().scrollPosition().isZero() || m_isQueued)
+        return;
+
+    if (!m_anchorElement || !m_anchorElement->renderBox()) {
+        chooseAnchorElement();
+        if (!m_anchorElement)
+            return;
+    }
+    m_isQueued = true;
+    m_frameView.queueAnchorControllerForUpdate(*this);
 }
 
 void ScrollAnchoringController::updateScrollPosition()
 {
-    // TODO: implement
-    // calling m_frame.scrollTo with different between m_anchorElement current position
-    // and m_lastPositionForAnchorElement
-    // should be called after suppression window (maybe Page::doAfterUpdateRendering())
+    SetForScope midAdjustingScrollPositionForAnchorElement(m_midAdjustingScrollPositionForAnchorElement, true);
+    auto queued = std::exchange(m_isQueued, false);
+    if (!m_anchorElement || !queued)
+        return;
+    auto renderBox = m_anchorElement->renderBox();
+    if (!renderBox) {
+        invalidateAnchorElement();
+        return;
+    }
+    FloatSize adjustment = computeOffset(*renderBox) - m_lastOffsetForAnchorElement;
+    if (!adjustment.isZero()) {
+        auto newScrollPosition = frameView().scrollPosition() + IntPoint(adjustment.width(), adjustment.height());
+        LOG(Scrolling, "ScrollAnchoringController::updateScrollPosition() scrolling to %d,%d", newScrollPosition.x(), newScrollPosition.y());
```

Comment:
This should log the frame in which the adjustment is happening. Easier to write with `LOG_WITH_STREAM()`

Reply:
Fixed.